quickjs.c

```diff
@@ -1774,10 +1774,27 @@ static int init_class_range(JSRuntime *rt, JSClassShortDef const *tab,
     return 0;
 }
 
-/* Note: OS and CPU dependent */
+/* Uses code from LLVM project. */
 static inline uintptr_t js_get_stack_pointer(void)
 {
-    return (uintptr_t)__builtin_frame_address(0);
+#ifdef _MSC_VER
+    return (uintptr_t) _AddressOfReturnAddress();
+#elif defined __has_builtin
+#if __has_builtin(__builtin_frame_address)
+    return (uintptr_t) __builtin_frame_address(0);
+#endif
+#elif defined __GNUC__
+    return (uintptr_t) __builtin_frame_address(0);
+#else
+    char CharOnStack = 0;
+    // The volatile store here is intended to escape the local variable, to
+    // prevent the compiler from optimizing CharOnStack into anything other
+    // than a char on the stack.
+    //
+    // Tested on: MSVC 2015 - 2019, GCC 4.9 - 9, Clang 3.2 - 9, ICC 13 - 19.
+    char *volatile Ptr = &CharOnStack;
+    return (uintptr_t) Ptr;
+#endif
```

Comment:
2 comments:
* please remove the space between the cast and its argument, as for all unary prefix operators.
* the case where `__has_builtin` is defined but `__has_builtin(__builtin_frame_address)` evaluates to false is not handled nor reported.  Why not use  
```
#elif defined __has_builtin && __has_builtin(__builtin_frame_address)
    return (uintptr_t)__builtin_frame_address(0);
#elif ...
```

Reply:
Good catch.
I am not sure though if macros are lazy and it won't fail if __has_builtin is not defined.
LLVM on the other hand does not even bother checking, so I guess this is more of a hypothetical scenario.